### ISSUE TITLE
Proposal to include pre-commit hook for python files

### ISSUE THREAD

**Issue by nico-stefani** [wazuh/wazuh#18802]
This issue aims to design a way to include a [pre-commit](https://pre-commit.com/) hook to work with the Python files.
The purpose is to unify rules for all developers involved in order to keep consistency in all our Python code and put the focus on the logic when we do PR reviews.

Some of the items that need to be reviewed, to include, are:
* Formatter
* Linter
* Security tool 
* etc

Also, we need to have a plan for how to handle the current code that is going to bring problems with the new rules.

**Comment by nico-stefani** [comment]
# Update

I found an interesting tool that includes linter (flake8, isort, bandit, etc) and formatter (black) capabilities, [Ruff](https://docs.astral.sh/ruff/). It's fast, configurable, and is monorepo-friendly. Also, it provides integrations with pre-commit and the more popular editors, such as VSCode, PyCharm, etc.

**Comment by nico-stefani** [comment]
I've been testing the tool and working on a POC configuration that matches our conventions.

<details><summary>.ruff.toml</summary>
<p>

```toml
# Include the Framework Team owned directories.
include = [
    "framework/**/*.py",
    "api/**/*.py",
    "wodles/**/*.py",
    "integrations/**/*.py",
]

# Same as Black.
line-length = 120
indent-width = 4

# Assume Python 3.8
target-version = "py37"

[lint]
# Enable Pyflakes (`F`) and a subset of the pycodestyle (`E`)  codes by default.
# Unlike Flake8, Ruff doesn't enable pycodestyle warnings (`W`) or
# McCabe complexity (`C901`) by default.
select = ["E4", "E7", "E9", "F", "I"]
ignore = ["F405"]

# Allow fix for all enabled rules (when `--fix`) is provided.
fixable = ["ALL"]
unfixable = []

# Allow unused variables when underscore-prefixed.
dummy-variable-rgx = "^(_+|(_+[a-zA-Z0-9_]*[a-zA-Z0-9]+?))$"

[per-file-ignores]
"wodles/**/*.py" = ["E402"]
"api/test/integration/env/configurations/**/*.py" = ["E402"]
"__init__.py" = ["F401"]

[format]
# Like Black, use double quotes for strings.
quote-style = "single"

# Like Black, indent with spaces, rather than tabs.
indent-style = "space"

# Like Black, respect magic trailing commas.
skip-magic-trailing-comma = false

# Like Black, automatically detect the appropriate line ending.
line-ending = "auto"
```

</p>
</details>

About the implementation strategy, I'll make sure to consult with the team.

**Comment by nico-stefani** [comment]
Continuing with POC, I added a configuration for the `pre-commit` tool

<details><summary>.pre-commit-config.yaml</summary>

```yaml
---
# See https://pre-commit.com for more information
# See https://pre-commit.com/hooks.html for more hooks
repos:
    - repo: https://github.com/astral-sh/ruff-pre-commit
      # Ruff version.
      rev: v0.1.7
      hooks:
          # Run the linter.
          - id: ruff
          # Run the formatter.
          - id: ruff-format
```
<p>

</p>
</details> 

And also, I've added a configuration sample for the `vs code` editor

<details><summary>VS Code configuration</summary>
<p>

```json
"[python]": {
        "editor.defaultFormatter": "charliermarsh.ruff",
        "editor.formatOnSave": false,
        "editor.formatOnType": true,
        "editor.codeActionsOnSave": {
            "source.fixAll": true,
            "source.organizeImports": false
        }
    }
```

</p>
</details> 

Regarding the implementation strategy, talking with the team we decided to point the changes to the next minor release(`4.9.0`), once the branch `4.8.0` is merged into master. Also, we decided to split the refactor into subtasks for each component(framework, API, wodles, integrations).

**Comment by GGP1** [comment]
## Review

The investigation was carried out successfully and obtained valuable results that would allow us to eliminate one of the steps that we are currently performing during the cross reviews. Furthermore. the corresponding issues were created to implement the changes necessary to get the pre-commit hook running.

I would like to add one more thing to the project and is the creation of a lint check that uses the same rules as the `.ruff.toml` file to verify that the pull request changes comply with them and that the developer's configuration is correct.

To conclude, I approve the investigation and think it should proceed to the final review.
